portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,44 +45,135 @@
 public class DataServlet extends HttpServlet {
 
   private final ArrayList<UserComment> userComments = new ArrayList<>();
+  private final String defaultMaxComment = "20";
 
   /*
    * Called when a client submits a GET request to the /data URL
    * Displays all recorded user comments on page
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int maxCommentDisplay = Integer.parseInt(getFieldFromResponse(request, "maxcomments", defaultMaxComment));
+    Query query = new Query("Comment").addSort("time", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    int displayed = 0;
+    ArrayList<UserComment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      if(displayed == maxCommentDisplay) {
+          break;
+      }
+      displayed++;
```

Comment:
Right now you
- Get all the comments from datastore
- Manually select the top-N comments

This might be inefficient (what if you have O(1000) of comments?).
Instead you could specify a limit for your query to control the maximum number of results returned in one batch:
https://cloud.google.com/datastore/docs/concepts/queries#cursors_limits_and_offsets

Reply:
I tried to use setLimit initially, but for some reason my code doesn't compile with it.

Reply:
It seems like the query they set us up with comes from the com.google.appengine.api.datastore.Query class but the above functions apply for the query from the com.google.cloud.datastore.Query<V> class.

Reply:
That is a good catch!

Seems like AppEngine datastore API is deprecated and people should in general use cloud datastore API: https://cloud.google.com/appengine/docs/standard/java/datastore

Is it possible to use cloud.datastore.query instead? It might break since the container is still AppEngine, but worth trying.

If we have to use AppEngine API, you can still setLimit
https://cloud.google.com/appengine/docs/standard/java/datastore/retrieving-query-results#setting_limits